cpp/src/arrow/ipc/message.cc

```diff
@@ -170,7 +170,7 @@ Status MaybeAlignMetadata(std::shared_ptr<Buffer>* metadata) {
   if (reinterpret_cast<uintptr_t>((*metadata)->data()) % 8 != 0) {
     // If the metadata memory is not aligned, we copy it here to avoid
     // potential UBSAN issues from Flatbuffers
-    RETURN_NOT_OK((*metadata)->Copy(0, (*metadata)->size(), metadata));
+    ARROW_ASSIGN_OR_RAISE(*metadata, (*metadata)->CopySlice(0, (*metadata)->size()));
```

Comment:
This seems like a common enough case, I think it'd be handy and easy to have a method `Copy()`:
```c++
Result<std::unique_ptr<Buffer>> Copy(MemoryPool* pool = default_memory_pool()) {
  return CopySlice(0, size(), pool);
}
```

Reply:
TBH I'm not sure copies of existing buffers are that common (most buffers are not mutated after being constructed).